### PR TITLE
Make it easier to build the dataset pipeline in Dataset C++ tests

### PR DESCRIPTION
**Pull request by feihugis**
This PR aims to simplify the code of making the dataset pipeline in Dataset C++ tests using the following style:
```C++
auto* source_dataset_params = new RangeDatasetParams(...);
auto* target_dataset_params = source_dataset_params->Batch(...)->Map(...)->...->...;
// Handles all the allocated dataset_params (source_dataset_params, batch_dataset_params, map_dataset_params, ... , target_dataset_params)
core::ScopedUnref scoped_unref(target_dataset_params); 
Tensor target_dataset;
MakeDatasetTensor(target_dataset_params, &target_dataset);
...
...

```
One example is as below
```C++
// The ownership is transferred to the caller.
MapDatasetParams* MapDatasetParams() {
  auto* range_dataset_params = new RangeDatasetParams(10, 0, -3);
  return range_dataset_params
      ->Batch(
          /*batch_size=*/2,
          /*drop_remainder=*/false,
          /*parallel_copy=*/true,
          /*output_dtypes=*/{DT_INT64},
          /*output_shapes=*/{PartialTensorShape({2})},
          /*node_name=*/"batch_node")
      ->Map(
          /*other_arguments=*/{},
          /*func=*/
          FunctionDefHelper::FunctionRef("XAddX", {{"T", DT_INT64}}),
          /*func_lib=*/{test::function::XAddX()},
          /*type_arguments=*/{},
          /*output_dtypes=*/{DT_INT64},
          /*output_shapes=*/{PartialTensorShape({1})},
          /*use_inter_op_parallelism=*/true,
          /*preserve_cardinality=*/false,
          /*node_name=*/kNodeName);
}
```

### CODE REVIEW

**Comment by feihugis** [comment]
@jsimsa The first commit (https://github.com/tensorflow/tensorflow/pull/31635/commits/79b55fd17b055c4a17737de87db6e0667685e2e5) is as same as #31592, which you have reviewed and is pending for merge. You can just review the second commit (https://github.com/tensorflow/tensorflow/pull/31635/commits/08ebb28d857e425134485487e3c65aeb38f0631c).

File: tensorflow/core/kernels/data/dataset_test_params.h [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+class UnaryDatasetParams : public DatasetParams {
+ public:
+  UnaryDatasetParams(DatasetParams* input_dataset_params,
+                     DataTypeVector output_dtypes,
+                     std::vector<PartialTensorShape> output_shapes,
+                     string node_name, DatasetParamsType type);
+
+  ~UnaryDatasetParams() {
+    if (input_dataset_params) {
+      input_dataset_params->Unref();
```
Comment (feihugis) [inline tensorflow/core/kernels/data/dataset_test_params.h]: Handles the release of input dataset params in the data pipeline.

File: tensorflow/core/kernels/data/dataset_test_params.h [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+                      std::vector<PartialTensorShape> output_shapes,
+                      string node_name, DatasetParamsType type);
+
+  ~BinaryDatasetParams() override {
+    if (input_dataset_params_0) {
+      input_dataset_params_0->Unref();
+    }
+    if (input_dataset_params_1) {
+      input_dataset_params_1->Unref();
+    }
```
Comment (feihugis) [inline tensorflow/core/kernels/data/dataset_test_params.h]: Handles the release of input dataset params in the data pipeline.

File: tensorflow/core/kernels/data/dataset_test_params.h [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+  ~DatasetParams() {}
+
+  // Returns the dataset inputs as a TensorValue vector.
+  virtual Status MakeInputs(gtl::InlinedVector<TensorValue, 4>* inputs) = 0;
+
+  // Checks if the tensor is a dataset variant tensor.
+  static bool IsDatasetTensor(const Tensor& tensor);
+
+  // Creates a BatchDatasetParams. The ownership is transferred to the caller.
+  BatchDatasetParams* Batch(int64 batch_size, bool drop_remainder,
```
Comment (feihugis) [inline tensorflow/core/kernels/data/dataset_test_params.h]: The ownership is transferred to the caller. That means the caller just needs to call `batch_dataset_params->UnRef()` or `core::ScopedUnref scoped_unref(batch_dataset_params)` to handle the deallocation for the whole pipeline.

Comment (feihugis) [reply tensorflow/core/kernels/data/dataset_test_params.h]: For example:
```C++
auto* range_dataset_params = new RangeDatasetParams(10, 0, -3);
auto* map_dataset_params = range_dataset_params->Batch(...)->Map(...);
core::ScopedUnref scoped_unref(map_dataset_params);
```
The allocated `range_dataset_params`, `batch_dataset_params`, and `map_dataset_params` will be automatically released.

File: tensorflow/core/framework/function_testlib.cc [hunk 4]
```diff
@@ -585,6 +585,76 @@ FunctionDef MakeRangeDataset() {
...
+        {"input_dataset", "other_arguments"},
+        {{"f", "$f"},
+         {"Targuments", "$Targuments"},
+         {"output_types", "$output_types"},
+         {"output_shapes", "$output_shapes"},
+         {"use_inter_op_parallelism", "$use_inter_op_parallelism"},
+         {"preserve_cardinality", "$preserve_cardinality"}}}});
+}
+
+FunctionDef MakeMapDatasetWithoutOtherArgs() {
```
Comment (jsimsa) [inline tensorflow/core/framework/function_testlib.cc]: there should be a single function with a boolean argument that controls `other_arguments` will be included

File: tensorflow/core/kernels/data/dataset_test_params.h [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+  Map,
+};
+
+string ToString(DatasetParamsType type);
+
+class RangeDatasetParams;
+class BatchDatasetParams;
+class MapDatasetParams;
+
+class DatasetParams : public core::RefCounted {
```
Comment (jsimsa) [inline tensorflow/core/kernels/data/dataset_test_params.h]: prefer using smart pointer over RefCounted ... but I don't think you need either ... you should be able to create static factories that return params object values and move these into subsequent factory invocations.

Comment (jsimsa) [reply tensorflow/core/kernels/data/dataset_test_params.h]: see my comment below

File: tensorflow/core/kernels/data/dataset_test_params.h (hunk 3 above)
Comment (jsimsa) [reply tensorflow/core/kernels/data/dataset_test_params.h]: Having these as member methods only works for unary transformations. However, for dataset sources (e.g. `range` or `random`) or binary (e.g. `concatenate`) or n-nary transformation (e.g. `zip`) this will not work.

Instead of a member method, there should be a static factory that takes the input dataset parameters as a parameter:

```
static DatasetParams Range(...);
static DatasetParams Batch(DatasetParams params, ...);
static DatasetParams Zip(std::vector<DatasetParams> params, ...);
```

and then use it as follows:

```
auto params = DatasetParams::CreateMap(std::move(DatasetParams::CreateBatch(std::move(DatasetParams::CreateRange(10, 0, -3)), ...)), ...);
```

File: tensorflow/core/kernels/data/dataset_test_params.h (hunk 1 above)
Comment (jsimsa) [reply tensorflow/core/kernels/data/dataset_test_params.h]: seeing as the ownership of this params object is never shared, I don't think you need refcounting or smart pointer and simply passing / moving the object values around will suffice

**Review by jsimsa — CHANGES_REQUESTED** [review]
Did a pass and left some comments.

File: tensorflow/core/kernels/data/dataset_test_params.h (hunk 3 above)
Comment (feihugis) [reply tensorflow/core/kernels/data/dataset_test_params.h]: @jsimsa Thanks for the prompt review! I have a few questions here.

>Instead of a member method, there should be a static factory that takes the input dataset parameters as a parameter

One potential issue for the static factory is that different static methods will have different input and return types as different DatasetParams have different attributes. For example:
```C++
static RangeDatasetParams Range(...);
static BatchDatasetParams Batch(T params, ...);
static ZipDatasetParams Zip(std::vector<DatasetParams> params, ...);
```
>`auto params = DatasetParams::Map(std::move(DatasetParams::Batch(std::move(DatasetParams::Range(10, 0, -3)), ...)), ...);`

The different types of input and return DatasetParams in each static method make it difficult to chain multiple dataset parameters together.

File: tensorflow/core/framework/function_testlib.cc (hunk 4 above)
Comment (feihugis) [reply tensorflow/core/framework/function_testlib.cc]: Yeah, will change that.

File: tensorflow/core/kernels/data/dataset_test_params.h (hunk 3 above)
Comment (jsimsa) [reply tensorflow/core/kernels/data/dataset_test_params.h]: Why can't they all return `DatasetParams` which the derived classes inherit from?

Comment (feihugis) [reply tensorflow/core/kernels/data/dataset_test_params.h]: I may misunderstand something. Here is an example:
```C++
static DatasetParams Range(int64 start, int64 stop, int64 step) {
  return RangeDatasetParams(start, stop, step);
}

DatasetParams range_dataset_params = Range(start, stop, step);
```
during the conversion, the values of `range_dataset_params.start`, `range_dataset_params.stop`, and `range_dataset_params.step` will be discarded.

**Comment by feihugis** [comment]
@jsimsa Thanks for your suggestion! Now `RefCounted` is replaced by the smart pointer and the static factory methods are added as below. Could you please have a look at the changes (https://github.com/tensorflow/tensorflow/pull/31635/commits/50624183a1e6c92546c268687afc6c6c817f8777) when you get a chance?
```C++
static std::shared_ptr<BatchDatasetParams> Batch(
    const std::shared_ptr<DatasetParams>& input_dataset_params, ...);

static std::shared_ptr<MapDatasetParams> Map(
    const std::shared_ptr<DatasetParams>& input_dataset_params, ...) ;

static std::shared_ptr<RangeDatasetParams> Range(int64 start, int64 stop, int64 step);
```

File: tensorflow/core/kernels/data/dataset_test_params.h [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+  // Returns the dataset inputs as a TensorValue vector.
+  virtual Status MakeInputs(gtl::InlinedVector<TensorValue, 4>* inputs) = 0;
+
+  // Checks if the tensor is a dataset variant tensor.
+  static bool IsDatasetTensor(const Tensor& tensor);
+
+  // Used to store all the input dataset parameters and the dataset tensors
+  // generated from the parameters.
+  std::vector<std::pair<std::shared_ptr<DatasetParams>, Tensor>>
+      input_dataset_params_group;
```
Comment (feihugis) [inline tensorflow/core/kernels/data/dataset_test_params.h]: `input_dataset_params_group` is used to handle n-nary transformation.

**Comment by mellanox-github** [comment]
Can one of the admins verify this patch?

**Review by jsimsa — COMMENTED** [review]
Thanks @feihugis did a pass and left a bunch of comments.

File: tensorflow/core/kernels/data/dataset_test_params.h [hunk 7]
```diff
@@ -139,102 +195,14 @@ class RangeDatasetParams : public DatasetParams {
...
-  ~RangeDatasetParams(){};
-
   Status MakeInputs(gtl::InlinedVector<TensorValue, 4>* inputs) override;
 
-  Tensor start;
-  Tensor stop;
-  Tensor step;
-};
-
-static std::shared_ptr<BatchDatasetParams> Batch(
```
Comment (feihugis) [inline tensorflow/core/kernels/data/dataset_test_params.h]: With the updated constructors of these DatasetParams class, the static factories are not needed.

File: tensorflow/core/kernels/data/batch_dataset_op_test.cc [hunk 8]
```diff
@@ -236,33 +212,43 @@ ITERATOR_GET_NEXT_TEST_P(BatchDatasetOpTest, BatchDatasetParams,
...
+                 /*parallel_copy=*/true,
+                 /*output_dtypes=*/{DT_INT64},
+                 /*output_shapes=*/{PartialTensorShape({4})})
+          .GetDatasetParams<BatchDatasetParams>();
   TF_ASSERT_OK(Initialize(&batch_dataset_params));
   TF_ASSERT_OK(CheckDatasetOutputDtypes({DT_INT64}));
 }
 
-std::vector<DatasetOutputShapesTestCase<BatchDatasetParams>>
+std::vector<DatasetOutputShapesTestCase<std::shared_ptr<BatchDatasetParams>>>
```
[the remark this answers is not in the document]
Comment (feihugis) [reply tensorflow/core/kernels/data/batch_dataset_op_test.cc]: `shared_ptr` is removed now.

File: tensorflow/core/kernels/data/dataset_test_params.cc [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+      node_name(std::move(node_name)),
+      type(type) {}
+
+bool DatasetParams::IsDatasetTensor(const Tensor& tensor) {
+  return tensor.dtype() == DT_VARIANT &&
+         TensorShapeUtils::IsScalar(tensor.shape());
+}
+
+BatchDatasetParams::BatchDatasetParams(
+    std::shared_ptr<DatasetParams> input_dataset_params, int64 batch_size,
```
[the remark this answers is not in the document]
Comment (feihugis) [reply tensorflow/core/kernels/data/dataset_test_params.cc]: After adding the implementation of `DatasetParams::MakeInputs`, the errors disappeared. But some members of the derived classes lose their value during the copy construction. One example can be found [here](https://coliru.stacked-crooked.com/view?id=ed75556c39348c9a) (the layer area should be same with the rectangle area).

The constructor is changed to be as below without using the smart pointer:
```C++
template <typename T>
  BatchDatasetParams(T input_dataset_params, int64 batch_size,
                     bool drop_remainder, bool parallel_copy,
                     DataTypeVector output_dtypes,
                     std::vector<PartialTensorShape> output_shapes,
                     string node_name)
```
Please feel free to let me know if you have any suggestion.

**Comment by feihugis** [comment]
@jsimsa Thanks for your review! The smart pointers are removed and all the members in DatasetParams are changed to be private. Could you please have a loot at the changes (https://github.com/tensorflow/tensorflow/pull/31635/commits/55afc02a147669a25e462abfafa684b9103476fa) when you get a chance?

File: tensorflow/core/framework/function_testlib.h [hunk 10]
```diff
@@ -40,6 +40,14 @@ class Attrs {
     }
   }
 
+  Attrs(
```
Comment (jsimsa) [inline tensorflow/core/framework/function_testlib.h]: why is this needed?

Comment (feihugis) [reply tensorflow/core/framework/function_testlib.h]: Some places like [here](https://github.com/tensorflow/tensorflow/pull/31635/files#diff-ba4d21ca0c437c902fd8c791e06ba44dR308) need it to convert [AttributeVector](https://github.com/tensorflow/tensorflow/pull/31635/files#diff-a4a3b8bfea4ac6f0185c2a132dadbb8bR32) to `Attrs`.

Comment (feihugis) [reply tensorflow/core/framework/function_testlib.h]: `AttributeVector` can not be replaced by `Attrs` as it is required by `test::function::NDef` like [here](https://github.com/tensorflow/tensorflow/pull/31635/files#diff-4085eb633f02b68967febffed95bc6edR41).

File: tensorflow/core/kernels/data/batch_dataset_op_test.cc [hunk 11]
```diff
@@ -109,96 +58,106 @@ class BatchDatasetOpTest : public DatasetOpsTestBaseV2<BatchDatasetParams> {
...
 BatchDatasetParams BatchDatasetParams3() {
-  return {/*num_input_elements=*/10,
-          /*batch_size=*/3,
-          /*drop_remainder=*/false,
-          /*parallel_copy=*/false,
-          /*output_dtypes=*/{DT_INT64},
-          /*output_shapes=*/{PartialTensorShape({-1})},
-          /*node_name=*/kNodeName};
+  auto range_dataset_params = RangeDatasetParams(0, 10, 1);
+  auto batch_dataset_params_0 =
```
Comment (jsimsa) [inline tensorflow/core/kernels/data/batch_dataset_op_test.cc]: no need to create a variable for this one, you can just return it

File: tensorflow/core/kernels/data/dataset_test_params.h [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+  std::vector<std::pair<std::shared_ptr<DatasetParams>, Tensor>>
+      input_dataset_params_group_;
+  DataTypeVector output_dtypes_;
+  std::vector<PartialTensorShape> output_shapes_;
+  string node_name_;
+  string iterator_prefix_ = kDefaultIteratorPrefix;
+  DatasetParamsType type_;
+};
+
+class BatchDatasetParams : public DatasetParams {
```
Comment (jsimsa) [inline tensorflow/core/kernels/data/dataset_test_params.h]: I feel like this should be in `batch_dataset_op_test.h`, inlining the implementation of `MakeInputs` and `MakeAttributes` so that the header can be included from test .cc files that wish to use it

Putting all params objects for all dataset op kernels into this file does not scale.

File: tensorflow/core/framework/function_testlib.h (hunk 10 above)
Comment (jsimsa) [reply tensorflow/core/framework/function_testlib.h]: ack

File: tensorflow/core/kernels/data/dataset_test_base.cc [hunk 13]
```diff
@@ -293,22 +292,61 @@ Status DatasetOpsTestBase::MakeRangeDataset(
   return Status::OK();
 }
 
-// Create a `RangeDataset` dataset as a variant tensor.
 Status DatasetOpsTestBase::MakeRangeDataset(
```
Comment (jsimsa) [inline tensorflow/core/kernels/data/dataset_test_base.cc]: similar to my comment about dataset_params, this should go into `range_dataset_op_test.h` ... putting all test dataset into this files does not scale

File: tensorflow/core/kernels/data/batch_dataset_op_test.cc (hunk 11 above)
Comment (feihugis) [reply tensorflow/core/kernels/data/batch_dataset_op_test.cc]: Yeah, changed it.

File: tensorflow/core/kernels/data/dataset_test_params.h (hunk 12 above)
Comment (feihugis) [reply tensorflow/core/kernels/data/dataset_test_params.h]: Great suggestion! Will change it as you suggest.

File: tensorflow/core/kernels/data/dataset_test_base.cc (hunk 13 above)
Comment (feihugis) [reply tensorflow/core/kernels/data/dataset_test_base.cc]: Yeah, I'm also planning to change this. After `DatasetParams` is updated with `MakeInputs()` and `MakeAttributes()`, we do not need to create a similar function for each dataset. A more general function can be used to replace them as below. 
```C++
Status DatasetOpsTestBase::MakeDatasetTensor(
      DatasetParams& dataset_params, 
      const FunctionDef& make_deataset_func, 
      Tensor* dataset_tensor);
```

**Comment by feihugis** [comment]
@jsimsa Thanks for your suggestions! The comments are addressed here(https://github.com/tensorflow/tensorflow/pull/31635/commits/7003529c6e7bdbc5ad6c493fe755ec8d2a88e44f). Also, clean up `DatasetOpsTestBaseV2` a little bit so that the derived test classes do not need to override the functions `Initialize()` and `MakeDatasetOpKernel`. Could you please have another look when you have time?

File: tensorflow/core/kernels/data/batch_dataset_op_test.cc [hunk 14]
```diff
@@ -11,194 +11,119 @@ limitations under the License.
...
 BatchDatasetParams BatchDatasetParams1() {
-  return {/*num_input_elements=*/12,
-          /*batch_size=*/4,
-          /*drop_remainder=*/false,
-          /*parallel_copy=*/true,
-          /*output_dtypes=*/{DT_INT64},
-          /*output_shapes=*/{PartialTensorShape({4})},
-          /*node_name=*/kNodeName};
+  auto range_dataset_params = RangeDatasetParams(0, 12, 1);
+  return BatchDatasetParams(std::move(range_dataset_params),
```
Comment (rachellim) [inline tensorflow/core/kernels/data/batch_dataset_op_test.cc]:
```suggestion
  return BatchDatasetParams(RangeDatasetParams(0, 12, 1),
```

File: tensorflow/core/kernels/data/batch_dataset_op_test.h [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+#ifndef TENSORFLOW_CORE_KERNELS_DATA_BATCH_DATASET_OP_TEST_H
+#define TENSORFLOW_CORE_KERNELS_DATA_BATCH_DATASET_OP_TEST_H
+
+#include "tensorflow/core/kernels/data/batch_dataset_op.h"
+#include "tensorflow/core/kernels/data/dataset_test_params.h"
+
+namespace tensorflow {
+namespace data {
+
+class BatchDatasetParams : public DatasetParams {
```
Comment (rachellim) [inline tensorflow/core/kernels/data/batch_dataset_op_test.h]: I'm in favor of having a dataset test utils file / namespace where these common classes are defined, instead of having a header file for batch dataset test specifically. It makes the test harder to read. Furthermore, you already have a bunch of utility-like files `dataset_test_{base | params}.{h | cc}`. I'd recommend coalescing all that into a single pair of `.h` and `.cc` files.

 I'll defer to @jsimsa on this though, if he thinks otherwise.

File: tensorflow/core/kernels/data/dataset_test_base.cc [hunk 16]
```diff
@@ -698,5 +678,100 @@ Status DatasetOpsTestBase::CheckIteratorSaveAndRestore(
   return Status::OK();
 }
 
+Status DatasetOpsTestBaseV2::Initialize(DatasetParams& dataset_params) {
```
Comment (rachellim) [inline tensorflow/core/kernels/data/dataset_test_base.cc]: While we're refactoring the tests, why do we have `DatasetOpsTestBase` and `DatasetOpsTestBaseV2`? Can you add comments that explain the distinction?

**Review by rachellim — COMMENTED** [review]
I haven't looked extensively at all of the changes here, but some comments to start.

File: tensorflow/core/kernels/data/dataset_test_base.cc (hunk 16 above)
Comment (feihugis) [reply tensorflow/core/kernels/data/dataset_test_base.cc]: Before this PR, `DatasetOpsTestBaseV2` will break the existing tests based on `DatasetOpsTestBase`. The changes in this PR will enable us to merge `DatasetOpsTestBaseV2` and `DatasetOpsTestBase`. Will do it once this PR is finalized.

File: tensorflow/core/kernels/data/batch_dataset_op_test.h (hunk 15 above)
Comment (feihugis) [reply tensorflow/core/kernels/data/batch_dataset_op_test.h]: There are two kinds of `DatasetParams`: 
1) the first one (e.g. `RangeDatasetParams`) will be used as the input dataset by other dataset op tests; 
2) the second kind of DatasetParams (e.g. `ReduceDatasetParams`) will be only used by itself. 

For the first one, we need to define it in some files (e.g. `dataset_test_params.h`) that can be imported by other tests; for the second one, we can define the dataset params class in its own test file (e.g. `reduce_dataset_op_test.cc`) as it will not be used by other tests. It may solve the scale issue that @jsimsa mentioned before.

`dataset_test_base.h` already has more than 600 lines of code, so I did not combine `dataset_test_base.{h | cc}` and `dataset_test_params.{h | cc}` together. If the length of the file is ok,  I will merge them into a single pair of `.h` and `.cc` file once @jsimsa and @rachellim make the final decision. 

Adding a namespace for the tests (e.g. `tensorflow::data::testing`) sounds great to me.

File: tensorflow/core/framework/function_testlib.cc [hunk 17]
```diff
@@ -585,6 +585,58 @@ FunctionDef MakeRangeDataset() {
...
+         {"output_types", "$output_types"},
+         {"output_shapes", "$output_shapes"}}}});
+}
+
+FunctionDef MakeMapDataset(bool has_other_args) {
+  auto args = has_other_args
+                  ? std::vector<string>{"input_dataset: variant",
+                                        "other_arguments: Targuments"}
+                  : std::vector<string>{"input_dataset: variant"};
+  auto input_placeholders =
```
[the remark this answers is not in the document]
Comment (feihugis) [reply tensorflow/core/framework/function_testlib.cc]: This name is not good. How about `node_args`?

File: tensorflow/core/framework/function_testlib.cc [hunk 18]
```diff
@@ -585,6 +585,58 @@ FunctionDef MakeRangeDataset() {
          {"output_shapes", "$output_shapes"}}}});
 }
 
+FunctionDef MakeBatchDataset() {
+  return FDH::Define(
+      // Name
```
[the remark this answers is not in the document]
Comment (feihugis) [reply tensorflow/core/framework/function_testlib.cc]: Thanks for the link!

File: tensorflow/core/kernels/data/batch_dataset_op_test.cc (hunk 14 above)
Comment (feihugis) [reply tensorflow/core/kernels/data/batch_dataset_op_test.cc]: If my understanding is correct, `std::move(range_dataset_params)` is used to avoid a copy here.

File: tensorflow/core/framework/function_testlib.cc (hunk 17 above)
[the remark this answers is not in the document]
Comment (rachellim) [reply tensorflow/core/framework/function_testlib.cc]: What about just `inputs`?

File: tensorflow/core/kernels/data/batch_dataset_op_test.cc (hunk 14 above)
Comment (rachellim) [reply tensorflow/core/kernels/data/batch_dataset_op_test.cc]: I believe the compiler will know to elide the copy for the temporary RangeDatasetParams object in this case -- https://en.cppreference.com/w/cpp/language/copy_elision

File: tensorflow/core/framework/function_testlib.cc (hunk 17 above)
[the remark this answers is not in the document]
Comment (feihugis) [reply tensorflow/core/framework/function_testlib.cc]: Yeah, sounds good. Will change it.

File: tensorflow/core/kernels/data/batch_dataset_op_test.cc (hunk 14 above)
Comment (feihugis) [reply tensorflow/core/kernels/data/batch_dataset_op_test.cc]: Yes, you are right. Will change it.

**Comment by feihugis** [comment]
Thanks @rachellim for your review and help! Most of the comments are addressed here (https://github.com/tensorflow/tensorflow/pull/31635/commits/015b45e95dac3047060f698b959083a4dcc6e2b4) except this one (https://github.com/tensorflow/tensorflow/pull/31635#discussion_r320406144). Will do it once we get the answer from @jsimsa.

File: tensorflow/core/kernels/data/batch_dataset_op_test.h (hunk 15 above)
Comment (jsimsa) [reply tensorflow/core/kernels/data/batch_dataset_op_test.h]: @feihugis if we only expect a small number of dataset parameter definitions to be shared, then @rachellim's suggestion makes sense.

I would not worry about the size of the `dataset_test_base` files as long as they are well organized and documented.

In summary, I support:
1) coalescing all test related functionality in `dataset_test_base.{h, cc}` putting shared dataset parameter definitions there and inlining all dataset parametr definitions in the `*_test.cc` files.

Comment (feihugis) [reply tensorflow/core/kernels/data/batch_dataset_op_test.h]: Great, will coalesce these `_params.h` files into  the `dataset_test_base` files.

**Comment by feihugis** [comment]
@rachellim All the comments are addressed now (https://github.com/tensorflow/tensorflow/pull/31635/commits/321e8cdf098d6a5b5645a9e3bbff0e0f6f223caf). Could you please take another look when you get a chance?

**Comment by feihugis** [comment]
> I'd suggest making everything consistent to _have_ a space after the colon, not removing the space from everything.

Gotcha. Will change them

**Comment by rachellim** [comment]
Thanks for your changes!

**Comment by feihugis** [comment]
Thanks for the very prompt review, @rachellim! This commit (https://github.com/tensorflow/tensorflow/pull/31635/commits/2bf79a761656e3c64fbaeead39e63e41fd492255) is submitted to address the comments. Please take a look!

**Review by jsimsa — DISMISSED** [review]

**Review by rachellim — DISMISSED** [review]

**Comment by feihugis** [comment]
The internal checks failed. Could you help check the logs and paste them here?

**Comment by jsimsa** [comment]
internal tests are still running

**Comment by feihugis** [comment]
@jsimsa The internal checks are in the failed status. Could you please help check the logs?

**Comment by jsimsa** [comment]
```
./third_party/tensorflow/core/kernels/data/dataset_test_base.h:132:3: error: 'tensorflow::data::DatasetParams' has virtual functions but non-virtual destructor [-Werror,-Wnon-virtual-dtor]
  ~DatasetParams() {}
  ^
./third_party/tensorflow/core/kernels/data/dataset_test_base.h:184:7: error: 'tensorflow::data::RangeDatasetParams' has virtual functions but non-virtual destructor [-Werror,-Wnon-virtual-dtor]
class RangeDatasetParams : public DatasetParams {
      ^
./third_party/tensorflow/core/kernels/data/dataset_test_base.h:208:7: error: 'tensorflow::data::BatchDatasetParams' has virtual functions but non-virtual destructor [-Werror,-Wnon-virtual-dtor]
class BatchDatasetParams : public DatasetParams {
      ^
./third_party/tensorflow/core/kernels/data/dataset_test_base.h:245:7: error: 'tensorflow::data::MapDatasetParams' has virtual functions but non-virtual destructor [-Werror,-Wnon-virtual-dtor]
class MapDatasetParams : public DatasetParams {
```

**Comment by feihugis** [comment]
@jsimsa The virtual deconstructor is added [here](https://github.com/tensorflow/tensorflow/pull/31635/commits/b0040a837c27bbd81e5f0033b4cd493a8f2513a9) and also update some places to use the named constant. Could you please take another look when you get a chance?

File: tensorflow/core/kernels/data/dataset_test_base.h [hunk 19]
```diff
@@ -129,7 +129,7 @@ class DatasetParams {
                 std::vector<PartialTensorShape> output_shapes, string node_name,
                 DatasetParamsType type);
 
-  ~DatasetParams() {}
+  virtual ~DatasetParams() {}
```
Comment (feihugis) [inline tensorflow/core/kernels/data/dataset_test_base.h]: Add the virtual destructor for the base class, but did not add the destructor for the derived class in order to make the code concise (also the destructor does not do anything here). But please feel free to let me know if it is a good practice to add the destructor for the derived classes.

**Review by jsimsa — APPROVED** [review]

**Comment by feihugis** [comment]
@rthadur An error happed while migrating the change. Could you help retrigger the internal checks? Thanks!

**Comment by rthadur** [comment]
done
